File: 卡码/代码随想录/C++/动态规划/dp/05-子序列问题/src/03-编辑距离/LeetCode_0115.cpp
//115. 不同的子序列
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
	int numDistinct(string s, string t) {
		uint dp[s.length()+1][t.length()+1];
		for(int i=0;i<s.length()+1;i++) dp[i][0]=1;
		for(int j=1;j<t.length()+1;j++) dp[0][j]=0;
		for(int i=1;i<s.length()+1;i++){
			for(int j=1;j<t.length()+1;j++){
				if(s[i-1]==t[j-1]) dp[i][j]=dp[i-1][j-1]+dp[i-1][j];
				else dp[i][j]=dp[i-1][j];
			}
		}
		return dp[s.length()][t.length()];
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/JS/动态规划/05-子序列问题/03-编辑距离/LeetCode_0115.js
//115. 不同的子序列
const numDistinct = function(s, t) {
	const dp=new Array(s.length+1);
	for(let i=0;i<s.length+1;i++){
		dp[i]=new Array(t.length+1).fill(0);
		dp[i][0]=1;
	}
	for(let i=1;i<s.length+1;i++){
		for(let j=1;j<t.length+1;j++){
			if(s[i-1]==t[j-1]) dp[i][j]=dp[i-1][j-1]+dp[i-1][j];
			else dp[i][j]=dp[i-1][j];
		}
	}
	return dp[s.length][t.length];
};
